File: knucleotide/knucleotide.go
/* The Computer Language Benchmarks Game
 * http://benchmarksgame.alioth.debian.org/
 *
 * contributed by The Go Authors.
 */

package main

import (
   "bufio"
   "bytes"
   "fmt"
   "io/ioutil"
   "os"
   "sort"
)

func count(data string, n int) map[string]int {
   counts := make(map[string]int)
   top := len(data) - n
   for i := 0; i <= top; i++ {
      s := data[i : i+n]
      counts[s]++
   }
   return counts
}

func countOne(data string, s string) int {
   return count(data, len(s))[s]
}

type kNuc struct {
   name  string
   count int
}

type kNucArray []kNuc

func (kn kNucArray) Len() int      { return len(kn) }
func (kn kNucArray) Swap(i, j int) { kn[i], kn[j] = kn[j], kn[i] }
func (kn kNucArray) Less(i, j int) bool {
   if kn[i].count == kn[j].count {
      return kn[i].name > kn[j].name // sort down
   }
   return kn[i].count > kn[j].count
}

func sortedArray(m map[string]int) kNucArray {
   kn := make(kNucArray, len(m))
   i := 0
   for k, v := range m {
      kn[i] = kNuc{k, v}
      i++
   }
   sort.Sort(kn)
   return kn
}

func printKnucs(a kNucArray) {
   sum := 0
   for _, kn := range a {
      sum += kn.count
   }
   for _, kn := range a {
      fmt.Printf("%s %.3f\n", kn.name, 100*float64(kn.count)/float64(sum))
   }
   fmt.Print("\n")
}

func main() {
   in := bufio.NewReader(os.Stdin)
   three := []byte(">THREE ")
   for {
      line, err := in.ReadSlice('\n')
      if err != nil {
         fmt.Fprintln(os.Stderr, "ReadLine err:", err)
         os.Exit(2)
      }
      if line[0] == '>' && bytes.Equal(line[0:len(three)], three) {
         break
      }
   }
   data, err := ioutil.ReadAll(in)
   if err != nil {
      fmt.Fprintln(os.Stderr, "ReadAll err:", err)
      os.Exit(2)
   }
   // delete the newlines and convert to upper case
   j := 0
   for i := 0; i < len(data); i++ {
      if data[i] != '\n' {
         data[j] = data[i] &^ ' ' // upper case
         j++
      }
   }
   str := string(data[0:j])

   var arr1, arr2 kNucArray
   countsdone := make(chan bool)
   go func() {
      arr1 = sortedArray(count(str, 1))
      countsdone <- true
   }()
   go func() {
      arr2 = sortedArray(count(str, 2))
      countsdone <- true
   }()

   interests := []string{"GGT", "GGTA", "GGTATT", "GGTATTTTAATT", "GGTATTTTAATTTATAGT"}
   results := make([]chan string, len(interests))
   for i, s := range interests {
      ch := make(chan string)
      results[i] = ch
      go func(result chan string, ss string) {
         result <- fmt.Sprintf("%d\t%s\n", countOne(str, ss), ss)
      }(ch, s)
   }
   <-countsdone
   <-countsdone
   printKnucs(arr1)
   printKnucs(arr2)
   for _, rc := range results {
      fmt.Print(<-rc)
   }

}

File: knucleotide/knucleotide.rb
# The Computer Language Benchmarks Game
# http://benchmarksgame.alioth.debian.org
#
# contributed by jose fco. gonzalez
# modified by Sokolov Yura

seq = ""

def frecuency(seq, length)
    n = seq.length - length + 1
    table = Hash.new(0)
    (0 ... length).each do |f|
        (f ... n).step(length) do |i|
            table[seq[i, length]] += 1
        end
    end
    [n, table]
end

def sort_by_freq(seq, length)
    n, table = frecuency(seq, length)
    table.to_a.sort{|a, b| b[1] <=> a[1]}.each do |v|
        puts "%s %.3f" % [v[0].upcase, ((v[1]*100).to_f / n)]
    end
    puts
end

def find_seq(seq, s)
    n, table = frecuency(seq, s.length)
    puts "#{table[s].to_s}\t#{s.upcase}"
end

start = false
STDIN.each_line do |line|
  start = true if line.start_with?(">THREE")

  if start && !line.start_with?(">")
    seq << line.chomp
  end
end

puts seq.length
(1..2).each { |i| sort_by_freq(seq, i) }
%w(ggt ggta ggtatt ggtattttaatt ggtattttaatttatagt).each { |s| find_seq(seq, s) }
